int32 nDevices
device[] devices 